locator/topology.hh

```diff
@@ -12,37 +12,128 @@
 
 #include <unordered_set>
 #include <unordered_map>
+#include <iostream>
 
 #include <seastar/core/future.hh>
 #include <seastar/core/sstring.hh>
 #include <seastar/util/bool_class.hh>
 
 #include "locator/types.hh"
 #include "inet_address_vectors.hh"
-#include "utils/fb_utilities.hh"
 
 using namespace seastar;
 
 namespace locator {
 
+class topology;
+
+class node : public enable_lw_shared_from_this<node> {
+public:
+    using local = bool_class<struct local_tag>;
+
+private:
+    host_id _host_id;
+    inet_address _endpoint;
+    endpoint_dc_rack _dc_rack;
+
+    // Private state of this instance.
+    // May change across topology generations
+    local _is_local;
+
+    friend class topology;
+public:
+    node(host_id id, inet_address endpoint, endpoint_dc_rack dc_rack, local is_local);
+
+    node(const node&) = default;
+    node(node&&) = delete;
+
+    const host_id& host_id() const noexcept {
+        return _host_id;
+    }
+
+    const inet_address& endpoint() const noexcept {
+        return _endpoint;
+    }
+
+    const endpoint_dc_rack& dc_rack() const noexcept {
+        return _dc_rack;
+    }
+
+    local is_local() const noexcept { return _is_local; }
+};
+
+using node_ptr = lw_shared_ptr<const node>;
+using node_set = std::unordered_set<node_ptr>;
```

Comment:
It's much better if this is an unordered_set<const node*> since manipulating it is less expensive that managing all those reference counts and conditionally calling destructors.

Reply:
Can be done as I said above, but it means that users must keep the topology valid while they need to use the node pointer since it will be invalidated on topology changes.

Reply:
removed in v8, will re-introduce on first use (in a followup PR)